airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/BigQueryDestination.kt

```diff
@@ -83,33 +98,217 @@ class BigQueryDestination : BaseConnector(), Destination {
                     .setUseLegacySql(false)
                     .build()
 
+            val result = BigQueryUtils.executeQuery(bigquery, queryConfig)
+
+            if (result.getLeft() == null) {
+                throw ConfigErrorException(result.right)
+            }
+
             if (UploadingMethod.GCS == uploadingMethod) {
-                val status = checkGcsPermission(config)
-                if (status!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
-                    return status
+
+                val GcsStatus = checkGcsAccessPermission(config)
+                if (GcsStatus!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
+                    return GcsStatus
                 }
+
+                //Copy a temporary dataset to confirm copy permissions are working
+                val bigQueryStatus = checkBigQueryCopyPermission(config);
+
+                if (bigQueryStatus!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
+                    return bigQueryStatus
+                }
+
             }
 
-            val result = BigQueryUtils.executeQuery(bigquery, queryConfig)
-            if (result.getLeft() != null) {
-                return AirbyteConnectionStatus()
-                    .withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+            return AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+
+        } catch (e: Exception) {
+            log.error(e) { "Check failed." }
+            throw ConfigErrorException((if (e.message != null) e.message else e.toString())!!)
+        }
+
+    }
+
+    /**
+     * This method performs a copy operation to copy data into a temporary table on BigQuery
+     * to check if the existing permissions are sufficient to copy data.
+     * If the permissions are not sufficient, then an exception is thrown with a message
+     * showing the missing permission
+     */
+    private fun checkBigQueryCopyPermission(config: JsonNode): AirbyteConnectionStatus? {
```

Comment:
avoid returning a nullable

Reply:
The existing methods such as "check" and "checkGcsAccessPermission" are also returning nullable. Is there any existing context about these methods returning nullable or should we make these methods to return non-nullable?

Reply:
We avoid nullable fields in Kotlin as much as possible. If the base class has a nullable in signature, continue to but the subsequent private methods avoid nullable.

Reply:
I have changed the code to remove the nullable